File: services/filesystem/write_text.ts
import { ensureParentDirectory } from "./ensure_parent.ts";
import type { WriteTextOptions } from "./types.ts";

export async function writeTextFile(
  path: string,
  content: string,
  options: WriteTextOptions = {},
): Promise<void> {
  await ensureParentDirectory(path);

  const final_content = options.ensure_trailing_newline === false
    ? content
    : content.endsWith("\n")
    ? content
    : `${content}\n`;

  await Deno.writeTextFile(path, final_content);
}

File: src/infrastructure/filesystem/write_text.rs
use std::fs;
use std::path::Path;

use super::ensure_parent_directory;
use crate::shared::FileSystemError;

pub fn write_text_file(
    path: &Path,
    content: &str,
    ensure_trailing_newline: bool,
) -> Result<(), FileSystemError> {
    ensure_parent_directory(path)?;

    let final_content = if !ensure_trailing_newline || content.ends_with('\n') {
        content.to_string()
    } else {
        format!("{}\n", content)
    };

    fs::write(path, final_content).map_err(|source| FileSystemError::WriteFile {
        path: path.to_path_buf(),
        source,
    })
}
